### PR TITLE
sanitize names from TS

### PR DESCRIPTION
**Pull request by Krovatkin**
This is to get a discussion started.
**Note,** the changes in `genInputDebugNames()` are to demonstrate that this fix covers the case with invalid input names to a fusion group as well, not just constants or intermediates. If we decide to go with this fix, I'll remove `genInputDebugNames` entirely. 

There seems to be two sources of invalid names:
1. TS where using "." is pretty common (for SSA naming and after freezing)
2. Duplicated names due to transformations in NNC. (https://github.com/pytorch/pytorch/issues/52727)
3. Technically, tensor expressions that are constructed with Python API could be another reason for invalid names, but the most of us have been conditioned to not use "." in names anyways, so this doesn't really seem to be an issue.

`1.` and `2.` seem orthogonal to me. I'd actually argue that we should be sanitizing names from TS as early as possible (maybe even before we create a kernel object) as this would result in consistent names throughout the entire NNC pipeline rather than doing it either after converting to the exprs to loop nests or in cuda codegen. 
Also, solving `1.` fixes the two test cases I added.

W.r.t. 2, I'm not sure how we should proceed here. It's not a functional bug in a sense it doesn't result in a compilation error or wrong result, rather in debuggability complications (see https://github.com/pytorch/pytorch/issues/52727) My intuition suggests that we should try keeping names consistent across the whole pipeline i.e. if we had a variable used in some loop and we peel an iteration from this loop, I'd rather the variable in the original loop to keep its name and the one in the peeled iteration to be renamed with some suffix.

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
Thanks for putting up a fix! I haven't thought about sanitizing on TS, but it seems like a good idea too!

> My intuition suggests that we should try keeping names consistent across the whole pipeline i.e. if we had a variable used in some loop and we peel an iteration from this loop, I'd rather the variable in the original loop to keep its name and the one in the peeled iteration to be renamed with some suffix.

The #52727 issue refers mostly to index variables which are not present in TS and only created as we lower to TE. For instance, a default name of the index variable is "i", thus most of the index variables in the TE IR currently have this name. When we print TE IR, we start appending "_NNN" to "i" for each new variable we see with this name, and the issue is that if we only print a part of the IR then the variable get different suffixes (as you can see in the example there).

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -3296,6 +3318,7 @@ void TensorExprKernel::compile() {
 
 TensorExprKernel::TensorExprKernel(const std::shared_ptr<Graph>& subgraph)
     : graph_(subgraph), code_(subgraph, "") {
+  sanitizeName(subgraph->block());
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/kernel.cpp:3321]: It would be better to call this from `compile()` just to keep all the transformations in one place.

**Review by navahgar — COMMENTED** [review]
The fix for 1 looks good to me. Just one minor nit below.

W.r.t 2., as @ZolotukhinM pointed out, the transformations we have pretty much do what you expected. If any of them do not, we should obviously fix it.

**Comment by ZolotukhinM** [comment]
On a technical side: what if in the JIT IR we had `%my_var` and `%my.var`. Wouldn't they be sanitized respectively to `%my_var` and `%my_var.1`?

**Comment by Krovatkin** [comment]
> On a technical side: what if in the JIT IR we had %my_var and %my.var. Wouldn't they be sanitized respectively to %my_var and %my_var.1?

Yes, they would be both sanitized to %my_var and then the `UniqueNameManager` will assign them different suffixes 😞 It looks like we will need to replace that `.` with `_` sooner or later before we hit `cuda_codegen`. I honestly think the "sooner" option will lead to the least confusion.

**Comment by ZolotukhinM** [comment]
> Yes, they would be both sanitized to %my_var

If we're doing it on the TorchScript IR, won't they get unique names there (i.e. won't the second one get the ".1" suffix there?)? If it will, then this solution won't be enough

**Comment by Krovatkin** [comment]
> If we're doing it on the TorchScript IR, won't they get unique names there (i.e. won't the second one get the ".1" suffix there?)? If it will, then this solution won't be enough

ah, that's correct. TS disambiguates on `setDebugName`, so we will get another more `.`s

**Comment by ZolotukhinM** [comment]
I think it's still a good idea to sanitize them on the TS level, maybe we can change that routine to use "_" instead of "."? And separately I think we could (and should) be able to sanitize and make names unique on the TE level - it could still be useful for scenarios where TE IR is constructed manually. But for the fuser use-case any of these two approaches would probably suffice.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 2]
```diff
@@ -2981,42 +3018,6 @@ TensorExprKernel::BackendType TensorExprKernel::inferBackendTypeFromDevice(
...
-  return sanitized_name.str();
-}
-
-// we use the debug names in printing cuda code, they need to be removed
-// of characters that can't be used in a variable identifier
-void TensorExprKernel::genInputDebugNames() {
-  std::unordered_map<std::string, const torch::jit::Value*> name_to_value;
-  std::unordered_set<std::string> name_set;
-  std::unordered_map<const torch::jit::Value*, std::string> value_to_name;
-  for (const torch::jit::Value* input : graph_->inputs()) {
```
Comment (eellison) [inline torch/csrc/jit/tensorexpr/kernel.cpp:3007]: can we just extend this to loop over constant nodes in the graph ? i would rather not modify the JIT IR, i think changing debug names and having them differ between the model graph and the optimized graph will hurt debugability

**Comment by Krovatkin** [comment]
@ZolotukhinM 
> And separately I think we could (and should) be able to sanitize and make names unique on the TE level - it could still be useful for scenarios where TE IR is constructed manually.

Yeah, in my initial fix, I added the sanitization logic in `Var` constructors, so we could cover the case where TE are constructed either with C++ or Python APIs, but then I realized that this case may not be too common and I was over-generalizing the fix.

**Maybe, we should actually do that.**

> I think it's still a good idea to sanitize them on the TS level, maybe we can change that routine to use "_" instead of "."? 

I could give it a try but I'd expect there would be too many little things to change across our tests.

What do you think of @eellison  suggestion to just extend his fix to constants?

I was also considering it. However, I wanted a bit more future proof fix in case there could be more things than just inputs, constants. @ZolotukhinM  @navahgar  what are yr thoughts, preferences?

**Comment by ZolotukhinM** [comment]
I think if we want to do it on the TS level, then your current approach is good - i.e. if we only limit it to constants then later it can break again if we add something else. If we don't want to touch TS (to address @eellison's concern on debuggability), then we should solve it on the TE side - but in this case I think the cleanest solution is to rename(make unique)+sanitize all variables in a stmt in one pass (as compared to just sanitizing names in Var constructor).

**Comment by Krovatkin** [comment]
>  (to address @eellison's concern on debuggability),

this is pretty subjective but we rarely need to follow a fusion group across both TS and NNC domains. Typically, it's either TS or NNC. If we sanitize as we enter the NNC domain, it shouldn't affect the debuggability too much as the names will be consistent across the NNC domain. just extra 2 cents here.

**Comment by ZolotukhinM** [comment]
> this is pretty subjective but we rarely need to follow a fusion group across both TS and NNC domains. Typically, it's either TS or NNC

Personally I also think that debuggability won't suffer much from your original solution, but I don't want to discount @eellison's opinion, esp. since we have an alternative solution here.

**Comment by eellison** [comment]
i'm good with whatever

**Pull request closed without merging**